core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -188,20 +185,13 @@ public void renameAllTempFiles() throws IOException {
     /**
      * Closes the current recovery target and waits up to a certain timeout for resources to be freed
      */
-    void resetRecovery(TimeValue timeout) throws IOException, TimeoutException {
+    void resetRecovery() throws InterruptedException {
         ensureRefCount();
         if (finished.compareAndSet(false, true)) {
```

Comment:
I think we should return a boolean to indicate whether this actually worked. I'm afraid that someone called the `fail` method concurrently, in which case the listener is notified and we can't reset the recovery any more. This means that if the caller of this methods get's a false back, it should cancel the new recovery.